contracts/Remittance.sol

```diff
@@ -1,16 +1,113 @@
+import "./utils/Pausable.sol";
 
-import {SafeMath} from "../node_modules/@openzeppelin/contracts/math/SafeMath.sol";
+import {SafeMath} from "./math/SafeMath.sol";
 
 contract Remittance is Ownable, Pausable {
 
-    using SafeMath for uint;
+    using SafeMath for uint256;
 
+    mapping(address => uint) public balances;
 
+    mapping(bytes32 => RemittanceInstance) public remittances;
+
+    struct RemittanceInstance {
+        address funder;
+        address broker;
+        uint256 fundsOwed;
+    }
+
+    event Deposit(
+        address indexed depositor,
+        uint256 amount
+    );
+
+    event RemittanceCreated(
+        bytes32 indexed hash,
+        address indexed funder,
+        address indexed broker,
+        uint256 amount
+    );
+
+    event RemittanceFundsReleased(
+        bytes32 indexed hash,
+        address indexed broker,
+        uint256 amount
+    );
+
+    event WithDraw(
+        address indexed withdrawer,
+        uint256 amount
+    );
+
+    // Generic deposit function
+    function deposit() external payable whenNotPaused {
+        require(msg.value > 0, "The value must be greater than 0");
+        balances[msg.sender] = balances[msg.sender].add(msg.value);
+
+        emit Deposit(msg.sender, msg.value);
+    }
+
+    // The funder can create a remittance
+    function create(bytes32 hash, address broker, uint256 amount)
+        external
+        whenNotPaused
+        returns(bool success)
+    {
+        require(broker != msg.sender, "The caller cannot be the broker");
+        require(amount > 0, "The amount must be greater than 0");
+
+        uint256 funderBalance = balances[msg.sender];
+
+        // Subtract from Funder balance
+        balances[msg.sender] = funderBalance.sub(amount);
+
+        RemittanceInstance storage remittanceInstance = remittances[hash];
+
+        remittanceInstance.funder = msg.sender;
+        remittanceInstance.broker = broker;
+        remittanceInstance.fundsOwed = amount;
+
+        emit RemittanceCreated(hash, msg.sender, broker, amount);
+
+        return true;
+    }
+
+    // The broker takes concatenates their password with the recipient's password to release the funds
+    function release(string memory concatenatedPassword) public whenNotPaused returns(bool success) {
+
+        require(bytes(concatenatedPassword).length > 0, "Password cannot be empty");
+        bytes32 hash = keccak256(abi.encodePacked(concatenatedPassword));
+
+        //Retrieve remittance
+        RemittanceInstance storage remittanceInstance = remittances[hash];
+
+        require(msg.sender == remittanceInstance.broker, "Only the broker can release funds");
+        require(remittanceInstance.fundsOwed > 0, "The funds have already been released");
+
+        emit RemittanceFundsReleased(hash, msg.sender, remittanceInstance.fundsOwed);
+        remittanceInstance.fundsOwed = 0;
+
+        (success, ) = msg.sender.call{value: remittanceInstance.fundsOwed}("");
```

Comment:
I'll make a bet: This has not been tested 😈

Reply:
Yes. I realise I have this omission. I had a poor memory of this comment in the splitter and consequently skipped this test.

![image](https://user-images.githubusercontent.com/2023203/108551884-77a1be80-72e8-11eb-9459-511e4f204e77.png)

After reading again, I have added a test to check that the contract's eth balance is checked after the funds have been released.

Is this adequate or do I also need to check the broker's eth account? I think the answer is yes.

Reply:
Well, the contract could have sent the funds to someone else, not the one you think it did.

Reply:
I have added:

```
        // Check the remittance amount has been sent to the broker ether balance
        const funderEthBalance = toBN(await web3.eth.getBalance(funder));
        const expectedFunderEthBalance = initFunderEthBalance.add(remittanceInstance.fundsOwed).toString(10);
        assert.strictEqual(funderEthBalance.toString(10), expectedFunderEthBalance);
```

This test passed. In the splitter, when a recipient calls withdraw(), they are charged gas. Why when the broker calls release() in this contract, and receives ether, are they not charged gas? This does not pass:

`const expectedFunderEthBalance = initFunderEthBalance.sub(cost).add(remittanceInstance.fundsOwed).toString(10);`

Reply:
Because this is not a matter of tx cost, it is a matter of sending `0`.
If you do `remittanceInstance.fundsOwed = 0` on line 92, then, on line 94, `value: remittanceInstance.fundsOwed` is essentially saying `value: 0`.

Reply:
:scream: :facepalm: You have reminded me that it is better to use explicit, hard coded values in my tests rather than variables as you suggested in the splitter tests.